cluster-api/pkg/controller/machineset/controller.go

```diff
@@ -0,0 +1,218 @@
+	}
+	defer c.queue.Done(key)
+
+	err := c.syncHandler(key.(string))
+	if err == nil {
+		c.queue.Forget(key)
+		return true
+	}
+
+	runtime.HandleError(fmt.Errorf("%v failed with: %v", key, err))
+	c.queue.AddRateLimited(key)
+
+	return true
+}
+
+// syncHandler holds the controller's business logic.
+// it makes sure that the current state is equal to the desired state.
+// note that the current state of the cluster is calculated based on the number of machines
+// that are owned by the given machineSet (key).
+func (c *Controller) syncHandler(key string) error {
+	machineSet, err := c.machineSetsLister.Get(key)
+	if err != nil {
+		return err
+	}
+
+	filteredMachines, err := c.getMachines(machineSet)
+	if err != nil {
+		return err
+	}
+
+	return c.syncReplicas(machineSet, filteredMachines)
+}
+
+// syncReplicas essentially scales machine resources up and down.
+func (c *Controller) syncReplicas(machineSet *v1alpha1.MachineSet, machines []*v1alpha1.Machine) error {
+	currentMachineCount := int32(len(machines))
+	desiredReplicas := int32(0)
+	if machineSet.Spec.Replicas != nil {
+	 desiredReplicas = *machineSet.Spec.Replicas
+	}
+
+	if desiredReplicas > currentMachineCount {
+		glog.V(2).Infof("creating a machine ( spec.replicas(%d) > currentMachineCount(%d) )", desiredReplicas, currentMachineCount)
+		machine, err := c.createMachine(machineSet)
+		if err != nil {
+			return err
+		}
+		_, err = c.machineClient.ClusterV1alpha1().Machines().Create(machine)
+		if err != nil {
+			return err
+		}
+		if err = waitForMachineInCache(machine.Name, c.machineLister); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	if desiredReplicas < currentMachineCount {
+		glog.V(2).Infof("deleting a machine ( spec.replicas(%d) < currentMachineCount(%d) )", desiredReplicas, currentMachineCount)
+		machineToDelete := machines[0].Name
+		err := c.machineClient.ClusterV1alpha1().Machines().Delete(machineToDelete, &metav1.DeleteOptions{})
+		if err != nil {
+			return err
+		}
+		if err = waitForMachineDeletedInCache(machineToDelete, c.machineLister); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	return nil
+}
+
+// createMachine creates a machine resource.
+// the name of the newly created resource is created as machineSet.Name + some random characters.
+// in addition, the newly created resource is owned by the given machineSet (we set the OwnerReferences field)
+func (c *Controller) createMachine(machineSet *clusterv1alpha1.MachineSet) (*clusterv1alpha1.Machine, error) {
+	machineName := machineSet.Name + "-" + string(uuid.NewUUID())[:6]
+
+	gv := clusterv1alpha1.SchemeGroupVersion
+	machine := &clusterv1alpha1.Machine{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            machineName,
+			OwnerReferences: []metav1.OwnerReference{*metav1.NewControllerRef(machineSet, gv.WithKind("MachineSet"))},
+		},
+		Spec: machineSet.Spec.Template.Spec,
+	}
+
+	return machine, nil
+}
+
+// getMachines returns a list of machines that match on machineSet.UID
```

Comment:
shouldn't this match on labels?

Reply:
I'm not sure whether we would like to make use of `labels` or not. For example the `ReplicaSet` controller makes use of `labels` to adopt orphaned `pods` (the ones that has been added manually). Nevertheless I would consider this as an advanced topic. Do you think that we could add an issue, so that we can discuss/work on that in the future ? Unless we consider this as something that blocks this PR.

Reply:
Don't use `OwnerReferences` - you want to remove a machine out of the `MachineSet`, do something with it, even delete the parent `MachineSet` without any impact of that machine. The `Selector` filed in the `MachineSet` is used exactly for this. Plus, querying for `Machines` that are part of your `MachineSet` will be easier.

Reply:
I am not against. But do we really want to remove `OwnerReferences` ? The biggest value they bring at the moment is automatic clean up (`GC`) of dependend objects. In fact that's why I haven't implemented `DeleteFunc` function as it was not clear what it could do?

This seems to be a good starting point as we have clear and very simple lifecycle management. In addition we mirror the default behaviour of `ReplicaSet` -  which deletes dependant object by default ?. I'm pretty sure that in the future we will add more strategies and policies around lifecycle management including the ability to remove `MachineSet` without removing dependent object so that they can be adopted by another `MachineSet`.

Reply:
We are trying to have the same experience and flow as in `Deployment` -> `ReplicaSet` -> `Pod`. Only the `Deployment` sets `OwnerReferences` to created `ReplicaSets`. `ReplicaSets` don't set `OwnerReferences` for `Pods`

Reply:
I have chatted with Martin and we have went through existing `ReplicaSet` code. It seems like `ReplicaSet` sets `OwnerReferences` on pods. 

More importantly during that discussion a suggestion came up that perhaps we should make an exact copy of an existing code (`ReplicaSet`) instead of crating our own controller from scratch. So what do you think ? 

a) we should start with something small and easy and build up on that? This essentially means that we could continue with this PR.

b) we should stop working on this PR and create a new one which an exact copy of `ReplicaSer` with its all functionality ?

Reply:
I think we could go both ways but I would lean towards starting with something small and building up on that so we can incorporate features as we are sure they make sense in machine management.

There might be differences between replica set and machine set that we need to consider. For example:
- In RS, there is garbage collection. What does garbage collection mean for machines? Are we going to have orphan machines that need to be collected? As I am not sure about how that would work, I'd consider not integrating that yet.
- There are properties that might be specific to pods, like burstReplicas, that we might need to think what that means for machines. We probably want to control any bursts too, but it might not be the same as bursting pods.
- In addition to those, creating and deleting machines might be more expensive and/or have consequences than pods. There are cases for clusters on premises that adding a machine is a very very long process (like filing a ticket for another team to provision a bare metal machine or go configure a VM on vSphere), which changes the characteristics of the problem.

In addition to that, I would personally prefer to add piece by piece and make sure they are well tested so it gives us time - as community - to start absorbing that code that we will own from this point on.

I hope that helps us move forward here, but also open to discuss further if others have any opinion about this.